plugins/openssl/openssl.cpp

```diff
@@ -0,0 +1,950 @@
+{
+    // Initial sanity check of our arguments
+    if (len_pem_public_keys == 0)
+        rtlFail(-1, "No public keys provided");
+
+    if (!isAll_pem_public_keys && len_plaintext > 0)
+    {
+        std::vector<EVP_PKEY *> publicKeys;
+        EVP_CIPHER_CTX * encryptCtx = nullptr;
+        byte ** encryptedKeys = nullptr;
+        MemoryBuffer iv;
+        MemoryBuffer ciphertext;
+
+        try
+        {
+            // Parse public keys and stuff them into a vector
+            const char * pubKeyPtr = static_cast<const char *>(_pem_public_keys);
+            const char * endPtr = pubKeyPtr + len_pem_public_keys;
+            while (pubKeyPtr < endPtr)
+            {
+                const size32_t keySize = *(reinterpret_cast<const size32_t *>(pubKeyPtr));
+                pubKeyPtr += sizeof(keySize);
+                publicKeys.push_back(pkeyCache.checkCache(keySize, pubKeyPtr, 0, nullptr));
+                pubKeyPtr += keySize;
+            }
+
+            // Load the cipher
+            const EVP_CIPHER * cipher = cipherCache.checkCache(_algorithm_name);
+
+            // Allocate memory for encrypted keys
+            size_t keyCount = publicKeys.size();
+            encryptedKeys = static_cast<byte **>(rtlMalloc(sizeof(byte *) * keyCount));
+            for (size_t x = 0; x < keyCount; x++)
+                encryptedKeys[x] = static_cast<byte *>(rtlMalloc(EVP_PKEY_size(publicKeys[x])));
+
+            // Allocate memory for the IV
+            int ivLen = EVP_CIPHER_iv_length(cipher);
+            iv.ensureCapacity(ivLen);
+
+            // Allocate buffer for ciphertext
+            int ciphertextLen = len_plaintext + EVP_CIPHER_block_size(cipher);
+            ciphertext.ensureCapacity(ciphertextLen);
+
+            // Create and initialize the context
+            encryptCtx = EVP_CIPHER_CTX_new();
+            if (!encryptCtx)
+                failOpenSSLError("creating cipher context");
+
+            // Initialize the envelope
+            std::vector<int> keyLens(keyCount);
+            if (EVP_SealInit(encryptCtx, cipher, encryptedKeys, keyLens.data(), static_cast<byte *>(iv.bufferBase()), publicKeys.data(), keyCount) != static_cast<int>(keyCount))
+                failOpenSSLError("EVP_SealInit");
+
+            // Update the envelope (encrypt the plaintext)
+            int len = 0;
+            if (EVP_SealUpdate(encryptCtx, static_cast<byte *>(ciphertext.bufferBase()), &len, reinterpret_cast<const byte *>(_plaintext), len_plaintext) != 1)
+                failOpenSSLError("EVP_SealUpdate");
+            ciphertextLen = len;
+
+            // Finalize the envelope's ciphertext
+            if (EVP_SealFinal(encryptCtx, static_cast<byte *>(ciphertext.bufferBase()) + len, &len) != 1)
+                failOpenSSLError("EVP_SealFinal");
+            ciphertextLen += len;
+
+            int totalKeyLen = 0;
```

Comment:
general: Use size32_t for the sizes of objects that are never going to be ridiculously large, size_t for full compatibility.  Same in various other places.